File: pyround1/374.猜数字大小.py
#
# @lc app=leetcode.cn id=374 lang=python
#
# [374] 猜数字大小
#

# @lc code=start
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
# def guess(num):

class Solution(object):
    def guessNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        l, r = 0, n
        while l <= r:
            m = (l + r) // 2
            res = guess(m)
            if res == 0:
                return m
            elif res == 1:
                l = m + 1
            else:
                r = m - 1
            
        
# @lc code=end

File: 374.猜数字大小.cpp
/*
 * @lc app=leetcode.cn id=374 lang=cpp
 *
 * [374] 猜数字大小
 */

// @lc code=start
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * int guess(int num);
 */

class Solution {
public:
    int guessNumber(int n) {
        if(!guess(0))
            return 0;
        if(!guess(n))
            return n ;
        long long start = 0;
        while(start < n){
            long long mid = (start + n) / 2;
            if(guess(mid) == 1)
            {
                start = mid;
            }
            else if(guess(mid) == -1)
            {
                n  = mid;
            }
            else
            {
                return mid;
            }
        }
        return n;
        
    }
};
// @lc code=end

